mysql_tests/test_crud.py

```diff
@@ -0,0 +1,333 @@
+    u = User(nickname="will-be-replaced", type=UserType.USER, age=42)
+    u.nickname = nickname
+    u.age = 21
+    await u.create(bind=engine, timeout=10)
+    assert u.id is not None
+    assert u.nickname == nickname
+    assert u.type == UserType.USER
+    assert u.age == 21
+
+    u2 = await User.get(u.id, bind=engine, timeout=10)
+    assert u2.id == u.id
+    assert u2.nickname == nickname
+    assert u2.type == UserType.USER
+    assert u2.age == 21
+    assert u2 is not u
+
+    return u
+
+
+async def test_get(engine):
+    u1 = await test_create(engine)
+    u2 = await User.get(u1.id, bind=engine, timeout=10)
+    assert u1.id == u2.id
+    assert u1.nickname == u2.nickname
+    assert u1 is not u2
+
+    u3 = await engine.first(u1.query)
+    assert u1.id == u3.id
+    assert u1.nickname == u3.nickname
+    assert u1 is not u3
+
+    u4 = await test_create_from_instance(engine)
+    u5 = await engine.first(u4.query)
+    assert u4.id == u5.id
+    assert u4.nickname == u5.nickname
+    assert u4 is not u5
+
+
+async def test_textual_sql(engine):
+    u1 = await test_create(engine)
+    u2 = await engine.first(
+        db.text("SELECT * FROM gino_users WHERE id = :uid")
+        .bindparams(uid=u1.id)
+        .columns(*User)
+        .execution_options(model=User)
+    )
+    assert isinstance(u2, User)
+    assert u1.id == u2.id
+    assert u1.nickname == u2.nickname
+    assert u1.type is u2.type
+    assert u1 is not u2
+
+    u2 = await engine.first(
+        db.text("SELECT * FROM gino_users WHERE id = :uid AND type = :utype")
+        .bindparams(db.bindparam("utype", type_=db.Enum(UserType)))
+        .bindparams(uid=u1.id, utype=UserType.USER,)
+        .columns(*User)
+        .execution_options(model=User)
+    )
+    assert isinstance(u2, User)
+    assert u1.id == u2.id
+    assert u1.nickname == u2.nickname
+    assert u1.type is u2.type
+    assert u1 is not u2
+
+
+async def test_select(engine):
+    u = await test_create(engine)
+    name = await engine.scalar(User.select("nickname").where(User.id == u.id))
+    assert u.nickname == name
+
+    name = await engine.scalar(u.select("nickname"))
+    assert u.nickname == name
+
+
+async def test_get_multiple_primary_key(engine):
+    u1 = await test_create(engine)
+    u2 = await test_create(engine)
+    await Friendship.create(bind=engine, my_id=u1.id, friend_id=u2.id)
+    with pytest.raises(ValueError, match="Incorrect number of values as primary key"):
+        await Friendship.get((u1.id,), bind=engine)
+    with pytest.raises(ValueError, match="Incorrect number of values as primary key"):
+        await Friendship.get(u1.id, bind=engine)
+    f = await Friendship.get((u1.id, u2.id), bind=engine)
+    assert f
+    assert f.my_id == u1.id
+    assert f.friend_id == u2.id
+
+
+async def test_multiple_primary_key_order():
+    import gino
+
+    db1 = gino.Gino()
+    await db1.set_bind(MYSQL_URL, autocommit=True)
```

Comment:
For `autocommit`: looks like this is a difference between asyncpg and DB-API. asyncpg has no concept of `autocommit` - any queries are executed in its own transaction and committed if there is no explicit transaction like a leading `BEGIN;` (similar to `autocommit=True` but SQLAlchemy is [removing autocommit in 2.0](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#migration-20-autocommit)). This is actually the nature of the [PostgreSQL binary protocol](https://segmentfault.com/a/1190000017136059#comment-area). As GINO < 1.1 is very asyncpg-specific, all APIs are assuming "auto-commit" - statements are simply "committed" when executed, unless they are wrapped in an explicit transaction.

I'm wondering if the MySQL protocol has similar behaviors or not - even though both psycopg2 and pymysql provided the (simulated) "non-auto-commit" feature. Either way, in the current GINO2 WIP, I'm using the ["driver-level auto-commit"](https://docs.sqlalchemy.org/en/14/changelog/migration_20.html#driver-level-autocommit-remains-available) to keep the GINO API be consistent with "auto-commit", so maybe we should do the same for MySQL in GINO 1.1, so that users don't have to specify `autocommit=True` for GINO to behave consistently.

Reply:
I haven't taken a closer look, but per [MySQL doc](https://dev.mysql.com/doc/refman/8.0/en/innodb-autocommit-commit-rollback.html), it's enabled by default. I'm not sure why it's not the case here. It can be caused by aiomysql.

I set it here just for easier testing.

Reply:
It's set to False by aiomysql. I think we can set it to None which means to use server default.

Reply:
Thanks for the fix! I'm drafting a doc about autocommit and SQLAlchemy 2.0, hopefully that could provide more ideas.

Reply:
Published here: https://python-gino.org/docs/en/master/explanation/sa20.html